src/Aspire.Hosting.Azure.AIFoundry/AzureAIFoundryLocalResourceExtensions.cs

```diff
@@ -0,0 +1,259 @@
+                        Properties = [.. state.Properties, new(CustomResourceKnownProperties.Source, "Foundry Local")]
+                    }).ConfigureAwait(false);
+                }
+
+            }, ct));
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Ensures that the resource stops when the application is shutting down.
+    /// </summary>
+    /// <param name="resource">The resource builder for the Foundry Local resource.</param>
+    /// <returns>The updated resource builder.</returns>
+    private static IResourceBuilder<AzureAIFoundryResource> EnsureResourceStops(this IResourceBuilder<AzureAIFoundryResource> resource)
+    {
+        resource.ApplicationBuilder.Eventing.Subscribe<ResourceReadyEvent>(resource.Resource, static (@event, ct) =>
+        {
+            _ = Task.Run(async () =>
+            {
+                var rns = @event.Services.GetRequiredService<ResourceNotificationService>();
+                var manager = @event.Services.GetRequiredService<FoundryLocalManager>();
+
+                await rns.WaitForResourceAsync(@event.Resource.Name, KnownResourceStates.Finished, ct).ConfigureAwait(false);
+
+                await manager.StopServiceAsync(ct).ConfigureAwait(false);
+            }, ct);
+
+            return Task.CompletedTask;
+        });
+
+        return resource;
+    }
+
+    /// <summary>
+    /// Configure a deployment for use with Foundry Local
+    /// </summary>
+    internal static IResourceBuilder<AzureAIFoundryDeploymentResource> AsLocalDeployment(this IResourceBuilder<AzureAIFoundryDeploymentResource> builder, AzureAIFoundryDeploymentResource deployment)
+    {
+        ArgumentNullException.ThrowIfNull(deployment, nameof(deployment));
+
+        var azureFoundryResource = builder.Resource.Parent;
+
+        builder.ApplicationBuilder.Eventing.Subscribe<ResourceReadyEvent>(azureFoundryResource, (@event, ct) =>
+        {
+            var rns = @event.Services.GetRequiredService<ResourceNotificationService>();
+            var loggerService = @event.Services.GetRequiredService<ResourceLoggerService>();
+            var logger = loggerService.GetLogger(deployment);
+            var manager = @event.Services.GetRequiredService<FoundryLocalManager>();
+            var eventing = @event.Services.GetRequiredService<IDistributedApplicationEventing>();
+
+            var model = deployment.ModelName;
+
+            _ = Task.Run(async () =>
+            {
+                await rns.PublishUpdateAsync(deployment, state => state with
+                {
+                    State = new ResourceStateSnapshot($"Downloading model {model}", KnownResourceStateStyles.Info),
+                    Properties = [.. state.Properties, new(CustomResourceKnownProperties.Source, model)]
+                }).ConfigureAwait(false);
+
+                var result = manager.DownloadModelWithProgressAsync(model, ct: ct) ?? throw new InvalidOperationException($"Failed to download model {model}.");
```

Comment:
When model download fails where will it show up? Who is logging this error?

Reply:
I'll have to check the latest iteration of the management SDK (we've got a snapshot of the code until their NuGet package ships) but I think it can't return `null` anymore, so it'll always return a result that we handle in the `await foreach` below.